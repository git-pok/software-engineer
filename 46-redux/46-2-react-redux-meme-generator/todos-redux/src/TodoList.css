.TodoList-ul {
  width: 100vw;
  height: 100vh;
  margin: 0px 0px;
  padding: 0px 0px;
  background-color: rgb(252, 252, 252);
  list-style: none;
}

.TodoList-container {
  display: flex;
  flex-direction: row;
  align-items: center;
  width: 100vw;
  padding: 0px 0px 0px 20px;
  margin: 20px 0px 0px 0px;
}

.TodoList-container li {
  padding: 0px 0px 0px 0px;
}

.TodoList-container button {
  margin: 0px 10px 0px 0px;
  width: 30px;
  height: 30px;
  text-align: center;
  border: 2px inset rgb(255, 255, 255);
  background-color: rgb(168, 170, 214);
  transition: ease-in-out .4s background-color;
}

.TodoList-true {
  text-decoration: line-through;
}

.TodoList-container button:hover {
  background-color: rgb(44, 43, 63);
  border-color: rgb(44, 43, 63);
  color: rgb(255, 255, 255);
}
