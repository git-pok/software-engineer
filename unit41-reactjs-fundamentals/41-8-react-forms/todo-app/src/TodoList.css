.TodoList {
  display: flex;
}


.TodoList-p {
  text-align: center;
}