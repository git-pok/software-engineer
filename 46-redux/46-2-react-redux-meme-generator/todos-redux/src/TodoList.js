import { useSelector, useDispatch } from "react-redux";
import './TodoList.css';

const TodoList = ({ createTodo }) => {

  const todoStore = useSelector(store => store.todos);
  const dispatch = useDispatch(store => store.todos);

  const deleteTodo = id => {
    dispatch({ type: "DELETE_TODO", payload: id });
  }

  const generateTodoHtml = todoStore.map(val => (
    <div className="TodoList-container" key={val.id}>
      <button onClick={() => deleteTodo(val.id)}>X</button>
      <li
        className={`TodoList-${val.status}`}>
          {val.todo}
      </li>
    </div>
  ))

  return (
    <ul className="TodoList-ul">
      {generateTodoHtml}
    </ul>
  );
}

export default TodoList;
