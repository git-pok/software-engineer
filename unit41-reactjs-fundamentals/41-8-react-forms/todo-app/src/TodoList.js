import { useState } from 'react';
import { v4 as uuid } from 'uuid';
import NewTodoForm from './NewTodoForm.js';
import Todo from './Todo.js';
import './TodoList.css';

/**
 * TodoList - this component should render the
 * NewTodoForm component and should render the list
 * of Todo components. Place your state that
 * contains all of the todos in this component.
 */

const TodoList = () => {
  
  const [ todos, setTodos ] = useState([]);

  const todosLen = todos.length;

  const addTodo = (newTodo) => {
    setTodos(todos => (
      [...todos, newTodo]
    ))
  }
  
  const removeTodo = (id) => {
    setTodos(todos.filter(todo => (
      todo.id !== id
    ))
    )
  }

  const editTodo = ({ id, edit }) => {
    setTodos(todos => (
      todos.map(todo => {
        if (todo.id === id) todo.todo = edit;
        return todo;
      })
    ))
  }

  return (
    <>
    <NewTodoForm addTodo={addTodo} />
      <div className="TodoList">
        <ul>
          { todosLen ?
            todos.map((list, idx) => (
             <li key={idx}><Todo
                                id={list.id}
                                todo={list.todo}
                                key={uuid()}
                                removeTodo={removeTodo}
                                editTodo={editTodo} /></li>
            )) :
            <h1 className="TodoList-p">No Todos</h1>
          }
        </ul>
      </div>
    </>
  );
}

export default TodoList;