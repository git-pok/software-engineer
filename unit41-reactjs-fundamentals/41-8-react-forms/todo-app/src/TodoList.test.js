import { render, fireEvent } from '@testing-library/react';
import {
  inputBc1, inputBc2, inputBc3,
  inputBcEdit1, inputBcEdit2
} from './_testCommon.js';
import TodoList from './TodoList.js';


test('renders <TodoList /> without crashing', () => {
  render(<TodoList />);
});


test('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});


test('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});


test('todo should not be rendered', () => {
  const { 
    queryByText
  } = render(<TodoList />);

  const box = queryByText("X");
  expect(box).not.toBeInTheDocument();
});


test('form adds new todo', () => {
  const { 
    getByText, getByLabelText, queryByText
  } = render(<TodoList />);

  const input = getByLabelText("Todo");
  const btn = getByText("ADD TODO");
  const noTodo = queryByText("Car Wash");
  expect(noTodo).not.toBeInTheDocument();

  fireEvent.change(input, inputBc1);
  fireEvent.click(btn);

  const todo = queryByText("Car Wash");
  const box = getByText("X");

  expect(todo).toBeInTheDocument();
  expect(box).toBeInTheDocument();
});


test('form inputs are empty after submit', () => {
  const { 
    getByText, getByLabelText
  } = render(<TodoList />);

  const input = getByLabelText("Todo");
  const btn = getByText("ADD TODO"); 

  fireEvent.change(input, inputBc1);

  fireEvent.click(btn);

  expect(input).toHaveValue("");
});


test('delete todo with X button', () => {
  const { 
    getByText, getByLabelText
  } = render(<TodoList />);

  const input = getByLabelText("Todo");
  const btn = getByText("ADD TODO"); 

  fireEvent.change(input, inputBc1);

  fireEvent.click(btn);

  const todoDeleteBtn = getByText("X");
  fireEvent.click(todoDeleteBtn);
  expect(todoDeleteBtn).not.toBeInTheDocument();
});


test('inputs should be empty', () => {
  const { 
    getByLabelText
  } = render(<TodoList />);

  const input = getByLabelText("Todo");

  expect(input).not.toHaveValue("Car Wash");
});


test('inputs should have values', () => {
  const { 
    getByLabelText
  } = render(<TodoList />);

  const input = getByLabelText("Todo");

  expect(input).not.toHaveValue("Car Wash");

  fireEvent.change(input, inputBc1);

  expect(input).toHaveValue("Car Wash");
});