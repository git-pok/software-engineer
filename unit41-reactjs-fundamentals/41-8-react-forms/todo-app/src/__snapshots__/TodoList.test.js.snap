// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewTodoForm-form"
  >
    <label
      class="NewTodoForm-label"
      for="todo"
    >
      Todo
    </label>
    <input
      class="NewTodoForm-input"
      id="todo"
      name="todo"
      placeholder="Create Todo!!!"
      type="text"
      value=""
    />
    <button
      class="NewTodoForm-submit"
    >
      ADD TODO
    </button>
  </form>
  <div
    class="TodoList"
  >
    <ul>
      <h1
        class="TodoList-p"
      >
        No Todos
      </h1>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 2`] = `
<DocumentFragment>
  <form
    class="NewTodoForm-form"
  >
    <label
      class="NewTodoForm-label"
      for="todo"
    >
      Todo
    </label>
    <input
      class="NewTodoForm-input"
      id="todo"
      name="todo"
      placeholder="Create Todo!!!"
      type="text"
      value=""
    />
    <button
      class="NewTodoForm-submit"
    >
      ADD TODO
    </button>
  </form>
  <div
    class="TodoList"
  >
    <ul>
      <h1
        class="TodoList-p"
      >
        No Todos
      </h1>
    </ul>
  </div>
</DocumentFragment>
`;
